tests/core/data_layer/test_data_rpc.py

```diff
@@ -132,6 +141,56 @@ async def test_create_insert_get(one_wallet_node_and_rpc: nodes_with_port, bt: B
         assert wallet_root["hash"] == bytes32([0] * 32)
         assert local_root["hash"] == None
 
+        # test empty changelist
+        changelist = []
+        with pytest.raises(ValueError, match="Changelist resulted in no change to tree data"):
+            await data_rpc_api.batch_update({"id": store_id.hex(), "changelist": changelist})
+
+
+@pytest.mark.asyncio
+async def test_upsert(one_wallet_node_and_rpc: nodes_with_port, bt: BlockTools) -> None:
+    wallet_node, full_node_api, wallet_rpc_port = one_wallet_node_and_rpc
+    num_blocks = 15
+    assert wallet_node.server
+    await wallet_node.server.start_client(PeerInfo("localhost", uint16(full_node_api.server._port)), None)
+    assert wallet_node.wallet_state_manager is not None
+    ph = await wallet_node.wallet_state_manager.main_wallet.get_new_puzzlehash()
+    for i in range(0, num_blocks):
+        await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph))
+        await asyncio.sleep(0.5)
+    funds = sum(
+        [calculate_pool_reward(uint32(i)) + calculate_base_farmer_reward(uint32(i)) for i in range(1, num_blocks)]
+    )
```

Comment:
This can stay for now since I'm still working on getting the test helpers into main then back here to atari in their new form...  but this is definitely not going to be the form going forward.